### PR TITLE
Functional: Modification to `where`

### PR DESCRIPTION
**Pull request by abishekg7**
## Description

Presently, the `where` clause in the example below yields a `float64` for `one_if_positive`, instead of a `Field[[EdgeDim, KDim], float]`, like we would expect in NumPy
```python
@field_operator
def _example_stencil(
    a: Field[[EdgeDim, KDim], float],
) -> Field[[EdgeDim, KDim], float]:
    one_if_positive = where(p_vn > 0.0, 1.0, 0.0)
```
The workaround, so far, has been to broadcast one of the floats into (EdgeDim, KDim) as shown below
```python
one_if_positive = where(p_vn > 0.0, broadcast(1.0, (EdgeDim, KDim)), 0.0)
```
This PR attempts to modify the type promotion behavior of `where` such that `where(a > 0.0, b, c)`, with `a` being a FieldType and `b` & `c` are ScalarType, yields a FieldType with the same dimensions as `a`.

### CODE REVIEW

File: src/functional/ffront/foast_passes/type_deduction.py [hunk 1]
```diff
@@ -509,6 +509,16 @@ def _visit_where(self, node: foast.Call, **kwargs) -> foast.Call:
...
+
+            if isinstance(mask_type, ct.FieldType):
+                if isinstance(return_type, ct.ScalarType):
+                    tmp_type = return_type
+                elif isinstance(return_type, ct.FieldType):
+                    tmp_type = return_type.dtype
+                return_type = type_info.promote(
+                    return_type, ct.FieldType(dims=mask_type.dims, dtype=tmp_type)
+                )
+
```
Comment (nfarabullini) [inline src/functional/ffront/foast_passes/type_deduction.py:521]: ` if isinstance(return_type, ct.ScalarType) and isinstance(mask_type, ct.FieldType):
                
return_type = type_info.promote(
                    return_type, ct.FieldType(dims=mask_type.dims, dtype=return_type)
                )`

**Review by nfarabullini — CHANGES_REQUESTED** [review]

File: src/functional/ffront/foast_passes/type_deduction.py (hunk 1 above)
Comment (abishekg7) [reply src/functional/ffront/foast_passes/type_deduction.py:517]: done. thanks!

File: src/functional/ffront/foast_passes/type_deduction.py [hunk 2]
```diff
@@ -509,6 +509,12 @@ def _visit_where(self, node: foast.Call, **kwargs) -> foast.Call:
 
         try:
             return_type = type_info.promote(left_type, right_type)
+
+            if isinstance(mask_type, ct.FieldType) and isinstance(return_type, ct.ScalarType):
+                return_type = type_info.promote(
+                    return_type, ct.FieldType(dims=mask_type.dims, dtype=return_type)
+                )
+
```
Comment (havogt) [inline src/functional/ffront/foast_passes/type_deduction.py:517]: just
```suggestion
            return_type = type_info.promote(mask_type, left_type, right_type)
```
should probably work

Comment (abishekg7) [reply src/functional/ffront/foast_passes/type_deduction.py:517]: That change results in this exception, probably as the dtype in `mask_type` is bool, and the others are float. 
```bash
   raise GTTypeError("Could not promote scalars of different dtype (not implemented).")
E   functional.common.GTTypeError: Could not promote scalars of different dtype (not implemented).
```

Comment (havogt) [reply src/functional/ffront/foast_passes/type_deduction.py:517]: oh that makes sense

File: tests/functional_tests/ffront_tests/test_execution.py [hunk 3]
```diff
@@ -609,6 +609,23 @@ def conditional_promotion(
...
+    def conditional_promotion(
+        mask: Field[[IDim], bool],
+    ) -> Field[[IDim], float64]:
+        return where(mask, 5.0, 10.0)
+
+    conditional_promotion(mask, out=out, offset_provider={})
+
+    assert np.allclose(np.where(mask, 5.0, 10), out)
+
+
```
Comment (havogt) [inline tests/functional_tests/ffront_tests/test_execution.py:628]: Lets add a test for the case with (IDim, JDIm) in the mask and (IDim) in the true/false branches. This probably does not work with the current implementation

Comment (abishekg7) [reply tests/functional_tests/ffront_tests/test_execution.py:628]: sure

Comment (abishekg7) [reply tests/functional_tests/ffront_tests/test_execution.py:645]: I added that test, and it failed. So I'm reverting back to my original logic for the type promotion. It seems to pass now.

File: src/functional/ffront/foast_passes/type_deduction.py [hunk 4]
```diff
@@ -509,6 +509,16 @@ def _visit_where(self, node: foast.Call, **kwargs) -> foast.Call:
 
         try:
             return_type = type_info.promote(left_type, right_type)
+
+            if isinstance(mask_type, ct.FieldType):
+                if isinstance(return_type, ct.ScalarType):
+                    tmp_type = return_type
```
Comment (havogt) [inline src/functional/ffront/foast_passes/type_deduction.py:515]:
```suggestion
                    return_dtype = return_type
```

File: src/functional/ffront/foast_passes/type_deduction.py [hunk 5]
```diff
@@ -509,6 +509,16 @@ def _visit_where(self, node: foast.Call, **kwargs) -> foast.Call:
...
+                elif isinstance(return_type, ct.FieldType):
+                    tmp_type = return_type.dtype
```
Comment (havogt) [inline src/functional/ffront/foast_passes/type_deduction.py:517]:
```suggestion
                    return_dtype = return_type.dtype
```

File: src/functional/ffront/foast_passes/type_deduction.py [hunk 6]
```diff
@@ -509,6 +509,16 @@ def _visit_where(self, node: foast.Call, **kwargs) -> foast.Call:
...
+                return_type = type_info.promote(
+                    return_type, ct.FieldType(dims=mask_type.dims, dtype=tmp_type)
```
Comment (havogt) [inline src/functional/ffront/foast_passes/type_deduction.py:519]:
```suggestion
                    return_type, ct.FieldType(dims=mask_type.dims, dtype=return_dtype)
```

**Review by havogt — CHANGES_REQUESTED** [review]
Let's improve readability a bit.

I am also not sure if we should introduce something like (pseudocode) `promoted_dims(field_or_scalar)->dims` as a helper.

**Comment by abishekg7** [comment]
@havogt Not sure re. helper function for `promoted_dims(field_or_scalar)->dims`. Would there be other situations requiring this you think?

File: src/functional/ffront/foast_passes/type_deduction.py (hunk 1 above)
Comment (abishekg7) [reply src/functional/ffront/foast_passes/type_deduction.py:521]: @nfarabullini had to revert your suggested change, due to an additional test failing. See Hannes' comment below.

**Review by havogt — APPROVED** [review]
Looks good to me

**Review by nfarabullini — APPROVED** [review]

**Pull request merged**